### PR TITLE
Waf configuration resource

### PR DESCRIPTION
**Pull request by oscarDovao**
This PR contains the implementation of the new waf version API endpoints.
Two functions provided for listing versions for a particular waf; one that takes parameter for pagination (pagination has to be done by the API's consumer) and another one which paginates through all the result set and returns the full list.

### CODE REVIEW

File: fastly/waf_version.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+	path := fmt.Sprintf("/waf/firewalls/%s/versions", i.WAFID)
+	err := c.paginateThroughLAllWAFVersions(path, i, r)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (c *Client) paginateThroughLAllWAFVersions(path string, i *ListWAFVersionsInput, r *WAFVersionResponse) error {
```
Comment (matpimenta) [inline fastly/waf_version.go]: There's maybe an extra `L` before `All`

File: fastly/waf_version.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+		return nil, err
+	}
+	return &wafVer, nil
+}
+
+// UpdateWAFInput is used as input to the UpdateWAFVersion function.
+type UpdateWAFVersionInput struct {
+	WAFID                            string
+	WAFVersion                       int
+	ID                               string `jsonapi:"primary,waf_firewall"`
```
Comment (matpimenta) [inline fastly/waf_version.go]: What's the different between `WAFID` and `ID`?

File: fastly/waf_version.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+// UpdateWAF updates a specific WAF version.
+func (c *Client) UpdateWAFVersion(i *UpdateWAFVersionInput) (*WAFVersion, error) {
+	if i.WAFID == "" {
+		return nil, ErrMissingWAFID
+	}
+
+	if i.WAFVersion == 0 {
+		return nil, ErrMissingWAFNumber
```
Comment (matpimenta) [inline fastly/waf_version.go]: Should this be `ErrMissingWAFVersion` since the check is against `WAFVersion`?

File: fastly/waf_version.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+// LockWAFVersion locks a specific WAF version.
+func (c *Client) LockWAFVersion(i *LockWAFVersionInput) (*WAFVersion, error) {
+	if i.WAFID == "" {
+		return nil, ErrMissingWAFID
+	}
+
+	if i.WAFVersion == 0 {
+		return nil, ErrMissingWAFNumber
```
Comment (matpimenta) [inline fastly/waf_version.go]: same here

File: fastly/waf_version_test.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
+package fastly
+
+import (
+	"reflect"
+	"strconv"
+	"testing"
+)
+
+func TestClient_WAF_Versions(t *testing.T) {
```
Comment (matpimenta) [inline fastly/waf_version_test.go]: Is there a way to split this test into small methods? It's quite long and difficult to follow

File: fastly/waf_version.go (hunk 1 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: well spotted, changed

File: fastly/waf_version.go (hunk 2 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: ID refers context id, in this case, WAFVersion id

File: fastly/waf_version.go (hunk 3 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: There seem to be 2 different terminologies here. *Version refers to the version number on the server side. if we want to be correct we can call it ErrMissingWAFVersionNumber

File: fastly/waf_version.go (hunk 4 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: see above

File: fastly/waf_version_test.go (hunk 5 above)
Comment (oscarDovao) [reply fastly/waf_version_test.go]: Restructure as discussed

**Pull request closed without merging**